addons/point_of_sale/models/pos_session.py

```diff
@@ -300,15 +305,485 @@ def _validate_session(self):
+                        # account
+                        line['income_account_id'],
+                        # sign
+                        -1 if line['amount'] < 0 else 1,
+                        # for taxes
+                        tuple((tax['id'], tax['account_id'], tax['tax_repartition_line_id']) for tax in line['taxes']),
+                    )
+                    sales[sale_key] = self._update_amounts(line['amount'], line['date_order'], **sales[sale_key])
+                    # Combine tax lines
+                    for tax in line['taxes']:
+                        tax_key = (tax['account_id'], tax['tax_repartition_line_id'], tax['id'], tuple(tax['tag_ids']))
+                        taxes[tax_key] = self._update_tax_amounts(tax['amount'], tax['base'], tax['date_order'], **taxes[tax_key])
+                # Combine stock lines
+                stock_moves = self.env['stock.move'].search([
+                    ('picking_id', '=', order.picking_id.id),
+                    ('company_id.anglo_saxon_accounting', '=', True),
+                    ('product_id.categ_id.property_valuation', '=', 'real_time')
+                ])
+                for move in stock_moves:
+                    exp_key = move.product_id.property_account_expense_id or move.product_id.categ_id.property_account_expense_categ_id
+                    out_key = move.product_id.categ_id.property_stock_account_output_categ_id
+                    amount = -sum(move.stock_valuation_layer_ids.mapped('value'))
+                    stock_expense[exp_key] = self._update_amounts(amount, move.picking_id.date, **stock_expense[exp_key])
+                    stock_output[out_key] = self._update_amounts(amount, move.picking_id.date, **stock_output[out_key])
+
+        ## SECTION: Create non-reconcilable move lines
+        # Create account.move.line records for
+        #   - sales
+        #   - taxes
+        #   - stock expense
+        #   - non-cash split receivables (not for automatic reconciliation)
+        #   - non-cash combine receivables (not for automatic reconciliation)
+        MoveLine = self.env['account.move.line'].with_context(check_move_validity=False)
+
+        tax_vals = [self._get_tax_vals(key, **amounts) for key, amounts in taxes.items()]
+        # Check if all taxes lines have account_id assigned. If not, there are repartition lines of the tax that have no account_id.
+        tax_names_no_account = [line['name'] for line in tax_vals if line['account_id'] == False]
+        if len(tax_names_no_account) > 0:
+            error_message = _(
+                'Unable to close and validate the session.\n'
+                'Please set corresponding tax account in each repartition line of the following taxes: \n%s'
+            ) % ', '.join(tax_names_no_account)
+            raise UserError(error_message)
+
+        MoveLine.create(
+            tax_vals
+            + [self._get_sale_vals(key, **amounts) for key, amounts in sales.items()]
```

Comment:
could you explicitly pass the arguments amounts['amount'] and amounts['amount_converted'] to clearly expose the structure. everywhere where **amounts is used